Singleton hack of fixing static initialisation order fiasco

Dalmurii opened the pull request:
https://github.com/llvm/llvm-project/issues/154528

# Brief
Indirect linking of llvm as a shared library is causing a "free() invalid size abortion". In my case, my project depends on google/clspv which in turn pulls `llvm`. Note that the issue does not occur when `clspv` and `llvm` is all statically linked.

# Structure of a project which might be causing an error
[google/clspv](https://github.com/google/clspv) has been depending on this project (llvm-project), as a static library.
My personal project has been depending on [google/clspv](https://github.com/google/clspv) as a shared library.
So `MyProject` was linked to shared object `clspv_core.so` which is containing `llvm-project` as its component.

# Problem
Linking `llvm-project` indirectly to `MyProject` via `clspv_core` was causing the `free() invalid size` abortion.
> When library is all statically linked, this problem did not occur.

[This issue](https://github.com/llvm/llvm-project/issues/154528) has a full log of the programme running with valgrind.

# Reason in my expectation
`KnownAssumptionStrings` from [clang/lib/Sema/SemaOpenMP.cpp](https://github.com/llvm/llvm-project/pull/154541/files#diff-032b46da5a8b94f6d8266072e296726c361066e32139024c86dcba5bf64960fc), [llvm/include/llvm/IR/Assumptions.h](https://github.com/llvm/llvm-project/pull/154541/files#diff-ebb09639e5957c2e4d27be9dcb1b1475da67d88db829d24ed8039f351a63ccff), [llvm/lib/IR/Assumptions.cpp](https://github.com/llvm/llvm-project/pull/154541/files#diff-1b490dd29304c875364871e35e1cc8e47bf71898affe3a4dbde6eb91c4016d06) and `FeatureMap` from [llvm/lib/Analysis/MLInlineAdvisor.cpp](https://github.com/llvm/llvm-project/pull/154541/files#diff-26c738eb291410ed83595a4162de617e8cbebddb46331f56d39d193868e29857), [llvm/include/llvm/Analysis/InlineModelFeatureMaps.h](https://github.com/llvm/llvm-project/pull/154541/files#diff-3b5a3359b2a0784186fb3f90dfabf905e8640b6adfd7d2c75259a6835751a6a7) which have been placed on global scope, causing static initialisation order ficasso when indirectly linked by `Myproject`.

# Fix trial
Changing those global instances I've mentioned ~ `KnownAssumptionStrings` and `FeatureMap`  ~ to functions which return a static variable's left value ~ `getKnownAssumptionStrings()`, `getFeatureMap()` ~  has solved my personal problem, so I am pulling a request of it.

Dalmurii committed:
Fix of [Static Initialization Order Fiasco Issue](https://github.com/llvm/llvm-project/issues/154528)

 Changes to be committed:
	modified:   clang/lib/Sema/SemaOpenMP.cpp
	modified:   llvm/include/llvm/Analysis/InlineModelFeatureMaps.h
	modified:   llvm/include/llvm/IR/Assumptions.h
	modified:   llvm/lib/Analysis/MLInlineAdvisor.cpp
	modified:   llvm/lib/IR/Assumptions.cpp

Dalmurii committed:
Merge remote-tracking branch 'ae2f' into HEAD

Dalmurii committed:
Merge branch 'llvm:main' into Singleton

Dalmurii committed:
clang-formatting the changes.

 Changes to be committed:
	modified:   llvm/include/llvm/Analysis/InlineModelFeatureMaps.h
	modified:   llvm/include/llvm/IR/Assumptions.h
	modified:   llvm/lib/Analysis/MLInlineAdvisor.cpp
	modified:   llvm/lib/IR/Assumptions.cpp

Dalmurii committed:
Merge branch 'llvm:main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii commented:
@mtrofin I have tried writing an explanation by editing the PR's first comment. Could you verify if this would explain enough or is it not enough for some reason?

mizvekov reviewed (approved):
LGTM, Thanks.

The changes are a strict improvement.

Pre-existing problem, but it sure looks iffy to me that these data structures are global, not attached to any context.

efriedma-quic commented on llvm/include/llvm/Analysis/InlineModelFeatureMaps.h:
```diff
@@ -160,7 +160,7 @@ inlineCostFeatureToMlFeature(InlineCostFeatureIndex Feature) {
 }
 
-LLVM_ABI extern std::vector<TensorSpec> FeatureMap;
+LLVM_ABI extern std::vector<TensorSpec> &getFeatureMap();
```
```suggestion
LLVM_ABI extern const std::vector<TensorSpec> &getFeatureMap();
```

efriedma-quic commented on llvm/include/llvm/IR/Assumptions.h:
```diff
@@ -30,18 +30,18 @@ constexpr StringRef AssumptionAttrKey = "llvm.assume";
 /// A set of known assumption strings that are accepted without warning and
 /// which can be recommended as typo correction.
-LLVM_ABI extern StringSet<> KnownAssumptionStrings;
+LLVM_ABI extern StringSet<> &getKnownAssumptionStrings();
```
```suggestion
LLVM_ABI extern const StringSet<> &getKnownAssumptionStrings();
```

Dalmurii committed:
Update llvm/include/llvm/Analysis/InlineModelFeatureMaps.h

> suggesting `getFeatureMap()` (past `FeatureMap`) could be constant.  
> suggesting `getKnownAssumptionStrings()` (past `KnownAssumptionStrings`) constant

Co-authored-by: Eli Friedman <efriedma@quicinc.com>

Dalmurii committed:
Update llvm/include/llvm/IR/Assumptions.h

> suggesting `getFeatureMap()` (past `FeatureMap`) could be constant.  
> suggesting `getKnownAssumptionStrings()` (past `KnownAssumptionStrings`) constant

Co-authored-by: Eli Friedman <efriedma@quicinc.com>

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii commented:
Felt not clear to force them to be constant though surely that changes could be solid when applicable. Resolving the issue.

Dalmurii commented:
Error has occurred. According to a log this says:
```
  /home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/include/llvm/IR/Assumptions.h:40:33: error: no matching member function for call to 'insert'
     40 |     getKnownAssumptionStrings().insert(AssumptionStr);
```

Clearly means `getKnownAssumptionStrings()` must be mutable. 
I will check if `getFeatureMap()` could be constant, and commit again.

Dalmurii commented:
According to a log, `getFeatureMap()` also requires to be mutable. Reverting last 2 commits I've made.

```
  /home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/lib/Analysis/MLInlineAdvisor.cpp:202:21: error: no matching member function for call to 'push_back'
    202 |     getFeatureMap().push_back(
```

Dalmurii committed:
Two functions requires to be mutable.

 On branch Singleton
 Changes to be committed:
	modified:   llvm/include/llvm/Analysis/InlineModelFeatureMaps.h
	modified:   llvm/include/llvm/IR/Assumptions.h

efriedma-quic commented:
That's broken: modifying global variables like that isn't threadsafe (e.g. if you're doing LTO).  But I guess it's equally broken with or without your patch.

Dalmurii commented:
> That's broken: modifying global variables like that isn't threadsafe (e.g. if you're doing LTO). But I guess it's equally broken with or without your patch.

Provided you say so, how do you think it should be resolved? Just making them constant would not work since somewhere it needs to be mutable.

efriedma-quic commented:
I think KnownAssumptionString is supposed to be some kind of plugin mechanism, but it's not a well-designed one.  It's exposing a global variable that modifies the frontend semantics, which is not how we want plugins to work; we have registration mechanisms for plugins which doesn't involve poking global variables. See https://reviews.llvm.org/D91979 .

For FeatureMap, I'm not sure if it's supposed to be a plugin mechanism, or it can just be a member variable MLInlineAdvisor.  The comments don't really explain why it's exposed in the header.

Dalmurii commented:
Opening another PR might be a proper fix.

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

Dalmurii committed:
Merge branch 'main' into Singleton

mtrofin commented on llvm/lib/Analysis/MLInlineAdvisor.cpp:
```diff
@@ -126,7 +130,7 @@ const TensorSpec llvm::DefaultDecisionSpec =
 const char *const llvm::RewardName = "delta_size";
 
-CallBase *getInlinableCS(Instruction &I) {
+static CallBase *getInlinableCS(Instruction &I) {
```
I'd separate the drive-by fix from the topic of this PR. You can do a follow-up "NFC" PR with just this. It adds an extra mental step for folks looking later at this change and trying to understand how this member static-ness relates to the topic of the PR.

mtrofin commented:
> p.s. The branch is saying it is out of date very frequently. Should I update branch or would it be fine to leave it?

Don't bother, not worth doing until the very end. If you don't have write permissions to the repo, I can help with the merging (to be clear, you maintain authorship, I can just push the button). If I hit any conflicts, I'll let you know.

mtrofin reviewed (approved):
lgtm, just please remove the unrelated fix in MLInlineAdvisor (the follow-up PR, if you want to do it, doesn't need anything stuffy - just label the title as "[NFC]" - because it's not changing any behavior - and just say something as simple as "made a member static in MLInlineAdvisor" or something like that.

Dalmurii committed:
Merge branch 'llvm:main' into Singleton

Dalmurii committed:
Removed primary changes to simplify the PR.

Dalmurii commented:
> lgtm, just please remove the unrelated fix in MLInlineAdvisor (the follow-up PR, if you want to do it, doesn't need anything stuffy - just label the title as "[NFC]" - because it's not changing any behavior - and just say something as simple as "made a member static in MLInlineAdvisor" or something like that.

Reverted making `CallBase *getInlinableCS(Instruction &I)` static on MLInlineAdvisor.

Removed from web viewer though seems it marks a change for some reason though style check is not saying anything.  
Would I need to clang-format it anyway?

mtrofin commented:
> For FeatureMap, I'm not sure if it's supposed to be a plugin mechanism, or it can just be a member variable MLInlineAdvisor. The comments don't really explain why it's exposed in the header.

@svkeerthy - @efriedma-quic is right, we should probably just make a local copy of the global FeatureMap and mutate that copy with the ir2vec extra features.

mtrofin committed:
Merge branch 'main' into Singleton

mtrofin commented:
> > lgtm, just please remove the unrelated fix in MLInlineAdvisor (the follow-up PR, if you want to do it, doesn't need anything stuffy - just label the title as "[NFC]" - because it's not changing any behavior - and just say something as simple as "made a member static in MLInlineAdvisor" or something like that.
> 
> Reverted making `CallBase *getInlinableCS(Instruction &I)` static on MLInlineAdvisor.
> 
> Removed from web viewer though seems it marks a change for some reason though style check is not saying anything. Would I need to clang-format it anyway?

Looks fine now.

mtrofin committed:
Merge branch 'main' into Singleton

mtrofin committed:
Merge branch 'main' into Singleton

mtrofin commented:
> That's broken: modifying global variables like that isn't threadsafe (e.g. if you're doing LTO). But I guess it's equally broken with or without your patch.
[...]
> For FeatureMap, I'm not sure if it's supposed to be a plugin mechanism, or it can just be a member variable MLInlineAdvisor. The comments don't really explain why it's exposed in the header.

Addressed in PR #156120. Thanks for the find!
